website/www/site/content/en/blog/scaling-streaming-workload.md

```diff
@@ -0,0 +1,291 @@
+
+After setting up the data generation and ingestion pipelines, and after letting the pipelines run for some minutes, we see that the pipeline is unable to sustain the desired throughput.
+
+<img class="center-block"
+    src="/images/blog/scaling-streaming-workload/1-default-ps.png"
+    alt="PubSub metrics">
+
+The previous image shows that the number of messages that are not being processed by the ingestion pipeline start to show as unacknowledged messages in Pub/Sub metrics.
+
+<img class="center-block"
+    src="/images/blog/scaling-streaming-workload/1-default-throughput.png"
+    alt="Throughput">
+
+Reviewing the per stage performance metrics, we see that the pipeline shows a saw-like shape, which is often associated with the throttling mechanisms the Dataflow runner uses when some of the stages are acting as bottlenecks for the throughput. Also, we see that the `Reshuffle` step on the `BigQueryIO` write transform does not scale as expected.
+
+This behavior happens because by default the [`BigQueryOptions`](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java#L57) uses 50 different keys to shuffle data to workers before the writes happen on BigQuery. To solve this problem, we can add a configuration to our launch script that enables the write operations to scale to a larger number of workers, which improves performance.
+
+
+## Second run: improve the write bottleneck
+
+After increasing the number of streaming keys to a higher number, 512 keys in our case, we restarted the test suite. The Pub/Sub metrics started to improve. After an initial ramp on the size of the backlog, the curve started to ease out.
+
+<img class="center-block"
+    src="/images/blog/scaling-streaming-workload/2-skeys-ps.png"
+    alt="PubSub metrics">
+
+This is good, but we should take a look at the throughput per stage numbers to understand if we are achieving the goal we set up for this exercise.
+
+<img class="center-block"
+    src="/images/blog/scaling-streaming-workload/2-skeys-throughput.png"
+    alt="Throughput">
+
+Although the performance has clearly improved, and the Pub/Sub backlog no longer increases monotonically, we are still far from the goal of processing one million events per second (1 GB/s) for our ingestion pipeline. In fact, the throughput metrics jump all over the time, indicating that bottlenecks are preventing the processing from scaling further.
+
+## Third run: unleash autoscale
+
+Luckily for us, when writing into BigQuery, we can autoscale the writes. This step simplifies the configuration so that we don't have to guess on the right number of shards. We switched the pipeline’s configuration and enabled this setting for the next [launch script](https://github.com/prodriguezdefino/apache-beam-streaming-tests/blob/main/scaling-streaming-workload-blog/3-ps2bq-si-tr-streamingautoshard.sh).
```

Comment:
```suggestion
Luckily for us, when writing into BigQuery, we can autoscale the writes. This step simplifies the configuration so that we don't have to guess the right number of shards. We switched the pipeline’s configuration and enabled this setting for the next [launch script](https://github.com/prodriguezdefino/apache-beam-streaming-tests/blob/main/scaling-streaming-workload-blog/3-ps2bq-si-tr-streamingautoshard.sh).
```